### ISSUE TITLE
[Feat] independent/fixed versioning with semantic-release

### ISSUE THREAD

**Issue by snebjorn** [nrwl/nx#2646]
I think it would be a great addition to be able to run semantic-release (or similar) on publishable libraries.

Lerna have something like this. I found this [mono repo](https://github.com/conventional-changelog/conventional-changelog) that demonstrates an independent setup

> lerna publish [--conventional-commits](https://github.com/lerna/lerna/blob/master/commands/version/README.md)

There's two modes to this, fixed and independent.

- **Fixed** will give each library the same version and have a CHANGELOG in the workspace root
- **Independent**  will give each library their own version and have a CHANGELOG in each project root

Hope you'll consider it

**Comment by Mani2693** [comment]
This will very useful expect this as soon as possible , Thanks !

**Comment by snebjorn** [comment]
Not sure what to do here. Should I keep this alive until it's implemented or rejected?

**Comment by Mani2693** [comment]
@vsavkin  Can you please help us to know when this will be taken for enhancement ?

**Comment by vsavkin** [comment]
Sorry. The issue was marked by mistake. I removed the label.

**Comment by iangregsondev** [comment]
Any movement on this or is it planned?

This is one key piece I am missing right now to be able to independently version each app in NX.

Anybody using any workarounds?

semantic-release is a great project but I would be willing to move to something else if its not possible.

If I could get each app independently versioned (package.json) and Tags in git - all the other stuff I want to do like docker deploys etc - would just fall into place :-)

I would be really interested in what others have done here - or is it just a non-starter?

**Comment by iangregsondev** [comment]
Any planned "deploy" commands for NX, I did check the community plugins and couldn't find anything that would work here - maybe I missed something?

**Comment by guiseek** [comment]
do we have something new with this work? is a very important feature, which is preventing large migrations to nx workspaces.

the community compodoc plugin seems to have solved it with an alternative solution, they created a release target in publishable projects.
you can see more here: https://github.com/twittwer/nx-tools/tree/master/tools/release

Do you believe it is a good way @vsavkin ?

**Comment by admir86** [comment]
It would be nice if it could also work for apps that are packaged in docker images. to version the docker images :)

**Comment by Sikora00** [comment]
I would love also a dedicated CHANGELOG for each application, and also an update of the version in package.json if exists in app.

**Comment by minijus** [comment]
Any chance to get this feature prioritized for version 11?

**Comment by glebmachine** [comment]
Will be nice to have!

**Comment by guiseek** [comment]
maybe present another solution with the community, like for example, what reason do we have to version all libraries together, I think it would be a great idea as long as we don't have a solution to version using semantic release

**Comment by rhys-e** [comment]
Couldn't you just add a `version` field to `workspace.json`? Then generate a CHANGELOG per lib/app.

**Comment by edbzn** [comment]
Hi there, we developed a community plugin for that [@jscutlery/semver](https://github.com/jscutlery/semver). It supports both modes **locked** and **fixed**. Take a look and give us some feedback.

**Comment by minijus** [comment]
There is quite fresh [@ng-builders/semrel](https://github.com/ng-builders/ng-builders/blob/master/libs/semrel.md#ng-builders-semrel-release) builder by @IKatsuba which looks promising as well. It might be worth a try!

**Comment by IKatsuba** [comment]
> There is quite fresh [@ng-builders/semrel](https://github.com/ng-builders/ng-builders/blob/master/libs/semrel.md#ng-builders-semrel-release) builder by @IKatsuba which looks promising as well. It might be worth a try!

This builder publishes itself!  I look forward to feedback and issues!)

**Comment by martaver** [comment]
@IKatsuba does this builder handle versioning for upstream non-lib dependents? E.g. if I have an angular site or a node app in my monorepo, will it version those? And if they depend on a lib that I published using this builder, will it also increment versions on the site and node app?

**Comment by IKatsuba** [comment]
@martaver you can create a feature request [here](https://github.com/ng-builders/ng-builders/blob/master/libs/semrel.md#ng-builders-semrel-release) and I will do it!)

**Comment by samuelfernandez** [comment]
Please keep this non stale, this would be much appreciated

**Comment by tingzuge-dio** [comment]
I am taking a naive cut at this in a plugin: https://github.com/abgov/nx-tools/blob/main/packages/nx-release/README.md.

```
npm i -D @abgov/nx-release
npx nx g @abgov/nx-release:lib my-lib
npx nx run my-lib:release
```
The generator creates semantic release configuration and the 'release' target just uses the run command. Next step is to add the executor and switch to js configuration with extension closer to nx.

Would love to have this 'out of the box'.

**Comment by dev-alex06** [comment]
Versioning with nx monorepo and commitizen would be nice. i am looking forward to a similar feature in the near future

**Comment by martaGalveSwipejobs** [comment]
It would be nice to get, if not an out of the box solution, some docs on potential workarounds.

**Comment by samuelfernandez** [comment]
@martaGalveSwipejobs Here you have a builder for that purpose as a 3rd party plugin, feedback is welcome! 😄 

https://github.com/ng-easy/platform/blob/main/libs/builders/README.md#semantic-release

https://www.npmjs.com/package/@ng-easy/builders

**Comment by MartaGalve** [comment]
@jscutlery/semver could be a good alternaive but it doesn't cover dependent libs, so if lib1 imports lib2 and lib2 changes, the version of lib1 is not updated. I haven't been able to find any library that covers this.

**Comment by samuelfernandez** [comment]
@MartaGalve [@ng-easy/builders](https://github.com/ng-easy/platform/blob/main/libs/builders/README.md#semantic-release) does just that. In the scenario you describe it'd do:

* Upgrade and release new version of lib2
* Detect that lib1 depends on lib2, so update it as well and release new version

The `@ng-easy/platform` monorepo's packages are being automatically built and released with that builder

**Comment by MartaGalve** [comment]
That's great @samuelfernandez, but @ng-easy/builders works only with Angular projects, doesn't it? In my case I'm working on a React/Nx project.

**Comment by samuelfernandez** [comment]
It should work with any Nx project, it is agnostic to what you are building 🚀

**Comment by sevakraynov** [comment]
@samuelfernandez 
Do you have some examples where used `@ng-easy/builders:semantic-release` in Nx?

**Comment by samuelfernandez** [comment]
@sevakraynov sure! The same builder is used to release itself 😁 source code here https://github.com/ng-easy/platform

**Comment by edbzn** [comment]
@MartaGalve A PR was open (https://github.com/jscutlery/semver/pull/278) to add the feature you talk about, you can give feedback it would help us.

**Comment by JoA-MoS** [comment]
Hey if you are still looking to some advice find me on the slack community (Justin Dietz) I publish semantic-release-plus which will work to have independent releases of publishable packages. https://www.npmjs.com/package/semantic-release-plus.

Examples of using it can be found here 
* https://github.com/semantic-release-plus/semantic-release-plus/tree/alpha

**Comment by broken-puppet** [comment]
I might be off the mark here but if you're trying to use conventional commits via Lerna alongside NX. [This](https://sonigaurav119.medium.com/how-i-use-lerna-with-nx-to-automate-the-versioning-1fada7f7f64f) article gives you a good workaround and also includes commitizen (if you want to use it).

I hope that helps.

**Comment by RS-Sautter** [comment]
ping, still needed, no?

**Comment by asilvadesigns** [comment]
👋

**Comment by RS-Sautter** [comment]
active
